kernel/mempool.c

```diff
@@ -301,6 +301,11 @@ int k_mem_pool_alloc(struct k_mem_pool *p, struct k_mem_block *block,
 	while (1) {
 		ret = pool_alloc(p, block, size);
 
+		/* [Sanechips] When race condition causes failure, re_alloc. */
```

Comment:
Why?  If the caller specified a timeout, this will loop already.  And if they want an instantaneous return, they're OK with dealing with an allocation failure in a graceful way.  I don't see a design reason to do extra work on their behalf -- let the caller decide what to do.